src/Container/ApplicationFactory.php

```diff
@@ -165,9 +166,20 @@ private function injectRoutes(Application $app, ContainerInterface $container)
             $methods = (isset($spec['allowed_methods']) && is_array($spec['allowed_methods']))
                 ? $spec['allowed_methods']
                 : null;
-            $name    = isset($spec['name']) ? $spec['name'] : null;
+            $name = isset($spec['name']) ? $spec['name'] : null;
             $methods = (null === $methods) ? Route::HTTP_METHOD_ANY : $methods;
-            $route   = new Route($spec['path'], $spec['middleware'], $methods, $name);
+
+            if (!is_callable($spec['middleware']) && is_array($spec['middleware'])) {
+                $middlewarePipe = new MiddlewarePipe();
+
+                foreach ($spec['middleware'] as $middleware) {
+                    $middlewarePipe->pipe(is_callable($middleware) ? $middleware : $container->get($middleware));
+                }
+
+                $spec['middleware'] = $middlewarePipe;
+            }
```

Comment:
Reviewing this, I wonder if it might make sense to do a "container-aware middleware pipe" which would lazy-load the middleware only when calling it. Looking into the Stratigility codebase, this would require both an alternate `Next` as well as `Dispatch` implementation — both of which would be easier to accomplish if we make a few properties protected and/or allow injection. I'll try to look into this this week.